pkg/utils/tcp/proxy.go

```diff
@@ -0,0 +1,217 @@
+package tcp
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"go.uber.org/zap"
+)
+
+type remote struct {
+	addr     string
+	mu       sync.Mutex
+	inactive bool
+}
+
+func (r *remote) isActive() bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return !r.inactive
+}
+
+func (r *remote) becomeInactive() {
+	r.mu.Lock()
+	r.inactive = true
+	r.mu.Unlock()
+	log.Debug("remote become inactive", zap.String("remote", r.addr))
+}
+
+func (r *remote) dial(timeout time.Duration) error {
+	conn, err := net.DialTimeout("tcp", r.addr, timeout)
+	if err != nil {
+		return err
+	}
+	conn.Close()
+	r.mu.Lock()
+	r.inactive = false
+	r.mu.Unlock()
+	return nil
+}
+
+type Proxy struct {
+	l             net.Listener
+	checkInterval time.Duration
+	dialTimeout   time.Duration
+	donec         chan struct{}
+	errc          chan error
+
+	mu        sync.Mutex
+	remotes   map[string]*remote
+	pickCount int // for RoundRobin count
+}
+
+func NewProxy(l net.Listener, endpoints map[string]string, checkInterval time.Duration, timeout time.Duration) *Proxy {
+	if checkInterval == 0 {
+		checkInterval = 5 * time.Second
+	}
+	if timeout == 0 {
+		timeout = 10 * time.Second
+	}
+	remotes := make(map[string]*remote)
+	for key, e := range endpoints {
+		remotes[key] = &remote{
+			addr:     e,
+			inactive: true,
+		}
+	}
+	return &Proxy{
+		l:             l,
+		errc:          make(chan error),
+		donec:         make(chan struct{}),
+		remotes:       remotes,
+		dialTimeout:   timeout,
+		checkInterval: checkInterval,
+	}
+}
+
+func (p *Proxy) updateRemotes(remotes map[string]string) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// remove all remotes
+	if remotes == nil {
+		p.remotes = make(map[string]*remote)
+		return
+	}
+	// update or create new remote
+	for key, nr := range remotes {
+		if cr, ok := p.remotes[key]; !ok {
+			log.Debug("proxy adds new remote", zap.String("remote", nr))
+			p.remotes[key] = &remote{
+				addr:     nr,
+				inactive: true,
+			}
+		} else if cr.addr != nr {
+			log.Debug("proxy updates existing remote", zap.String("old", cr.addr), zap.String("new", nr))
+			cr.addr = nr // could cause data race but doesnt matter
+			cr.becomeInactive()
+		}
+	}
+	// remove old remote
+	for key, r := range p.remotes {
+		if _, ok := remotes[key]; !ok {
+			log.Debug("proxy discards remote", zap.String("remote", r.addr))
+			delete(p.remotes, key)
+		}
+	}
+}
+
+func (p *Proxy) serve(in net.Conn) {
+	var (
+		err    error
+		out    net.Conn
+		picked *remote
+	)
+	for {
+		p.mu.Lock()
+		picked = p.pick()
+		p.mu.Unlock()
```

Comment:
some mutexs are placed in the function definition while some are placed at call site, prefer to unify them.